### ISSUE TITLE
Adopt SWR instead of data-provider?

### ISSUE THREAD

**Issue by bryceosterhaus** [liferay/clay#3465]
I was looking into https://swr.vercel.app/ recently and was playing around with the idea in my head of just adopting that as our data fetching tool rather than maintain our own.

Any thoughts on this? From a quick overview, SWR seems to cover most if not all use cases we provide with data-provider. SWR is even smaller than data-provider according to bundle phobia.

[SWR Bundle Phobia](https://bundlephobia.com/result?p=swr@0.2.3)

cc @matuzalemsteles

**Comment by matuzalemsteles** [comment]
Hey @bryceosterhaus, I feel that SWR has a different strategy than we have. It seems to follow a very strong strategy in `stale-while-revalidate`, apparently, you do not have an option to switch this cache strategy, so you will always take it from the cache and revalidate it on the server, I feel that this is not always a good strategy. I would say that we are not far behind a customer foundation, the SWR seems to have quite a few features geared towards revalidating the cache. If we look at resource comparison we have the following:

|Features|**useResource**|**SWR**| 
|---|---|---|
| Cache |✅|✅|
| Fetch Policy |✅|🚫|
| Cache Strategy |LRU|?|
| Dedupe requests |⚠️|✅|
| Retry |✅ 🆙|✅|
| Network Status |✅ 🆙|✅|
| Revalidate on Focus |🚫|✅|
| Revalidate on Reconnect |🚫|✅|
| Revalidate on Mount |?|✅|
| Polling |✅|✅ 🆙|
| Suspense |⚠️|✅|
| Custom Fetch |⚠️|✅|
| Data dependency | ⚠️ | ✅|

- ✅ Done
- 🚫 Not implemented
- ✅ 🆙 Better configs
- ⚠️ Implementation in progress or roadmap

Looking closely, there are still some missing resources but the vast majority already have PRs on hold and some forwarded for discussion. There are still other features I want to implement with #2782, apparently, you can do this with `mutation` in the SWR but it wasn’t very clear to me.

We also don't have revalidation implementations, which are UI and network-oriented, but some of them may not make sense for the data-provider... 🤔

In summary, I like the features of SWR but only the cache strategy does not please me so much and the possibility of not configuring. Perhaps the only advantage really would be to reduce maintenance over the data provider, but I think that at the time we will have to support the Data Provider until some point. There is also [React Query](https://github.com/tannerlinsley/react-query) if we want to evaluate the options.

**Comment by jbalsas** [comment]
Up to you guys!

> Perhaps the only advantage really would be to reduce maintenance over the data provider

This is juicy... comes at the cost of additional dependencies... but can also make code less "proprietary"... I'm good with whatever choice you make 👌

**Comment by bryceosterhaus** [comment]
> It's important to focus on our "core business"

This might be more of the direction I am thinking. Every time I think about data-provider in the context of "Clay" I'm not quite sure what need we are exactly fulfilling with it. I did just search through portal and it is used 11 times, so we do use it a bit and it must be meeting a need. I think I would just be confused as a user seeing it next to a bunch of ui components and it being data fetching focused.

**Comment by matuzalemsteles** [comment]
> This might be more of the direction I am thinking. Every time I think about data-provider in the context of "Clay" I'm not quite sure what need we are exactly fulfilling with it. I did just search through portal and it is used 11 times, so we do use it a bit and it must be meeting a need. I think I would just be confused as a user seeing it next to a bunch of ui components and it being data fetching focused.

That's a good point, sorry for the delay. I'm thinking about it, I don't know if the SWR would be best to cover our use cases today, so I don't know if we should recommend it yet. I'm thinking of moving the Data Provider to the DXP codebase and keeping it there, probably there makes more sense then? @wincent what do you think?

**Comment by jbalsas** [comment]
> I'm thinking of moving the Data Provider to the DXP codebase and keeping it there, probably there makes more sense then?

I was thinking about this the other day... glad you brought this up!

I think moving this to DXP for now is a good solution... in Clay, we leave the decision up to the developer, and in DXP we offer a more opinionated way to fetch resources. It would also help us more clearly separate the APIs from our implementation and make components that can compose better...

I'll leave the decision up to you, but you have my 👍

**Comment by wincent** [comment]
> @wincent what do you think?

- I'm fine with moving it to DXP. It doesn't seem "core" Clay.
- I agree we shouldn't race to adopt SWR (or anything else) yet; decisions like that tend to have long-term consequences so should be taken very carefully after a reasonable amount of hands-on experimentation, testing and analysis  — no change at all is the course of least risk for now (in some situations, inaction is in itself a risk, but I don't think we're anywhere near that point yet).

**Comment by matuzalemsteles** [comment]
Oh, this is great! good to hear from you, I think this will be the best way now. We can move a part of the documents to the guidelines maybe, and move the source code to `frontend-js-react-web` and we can modify some things like directly using `fetch` from `frontend-js-web`... @bryceosterhaus what do you think?

**Comment by bryceosterhaus** [comment]
Yeah that sounds good to me if we move data-provider to dxp. I think it makes more sense in that context and that way we can decouple it with clays's ui focus. I mainly threw swr out there as an idea, I just thought it looked interesting.

Thanks for the feedback everyone

**Comment by matuzalemsteles** [comment]
Great! So, we can walk with it. We can create tickets to go with this, but as we have no prey on that, we can create them next week.
